### PR TITLE
SDK-30 [Java SDK] use oAuth protocol for PoPAPI communication

### PR DESCRIPTION
**Pull request by VladimirLevitskiy**
Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- [SDK-30](https://incountry.atlassian.net/browse/SDK-30)

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [ ] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [ ] Please put the target release vehicle this needs to be bundled with.
- [ ] Did you increment the version? What is the new version?

### CODE REVIEW

File: src/main/java/com/incountry/residence/sdk/StorageImpl.java [hunk 1]
```diff
@@ -117,53 +127,65 @@ public static Storage getInstance(String environmentID, String apiKey, String en
...
-        instance.encrypted = secretKeyAccessor != null;
-        return instance;
+        instance.dao = initDao(config, dao);
+        instance.encrypted = config.getSecretKeyAccessor() != null;
+        instance.cryptoManager = new CryptoManager(config.getSecretKeyAccessor(), config.getEnvId(), config.getCustomEncryptionConfigsList(), config.isNormalizeKeys());
+        return ProxyUtils.createLoggingProxyForPublicMethods(instance);
+    }
+
+    private static Dao initDao(StorageConfig config, Dao dao) throws StorageServerException, StorageClientException {
+        if (dao == null) {
```
[the remark this answers is not in the document]
Comment (VladimirLevitskiy) [reply src/main/java/com/incountry/residence/sdk/StorageImpl.java:168]: This constructor is used for setting mocks for unit tests

File: src/main/java/com/incountry/residence/sdk/tools/dao/impl/HttpDaoImpl.java [hunk 2]
```diff
@@ -104,34 +107,38 @@ private String getEndpoint(String path, String country) throws StorageClientExce
...
     private String createUrl(String country, String keyHash) throws StorageClientException, StorageServerException {
-        return getEndpoint(concatUrl(country, URI_DELIMITER, keyHash), country);
+        return getEndpoint(country) + concatUrl(country, URI_DELIMITER, keyHash);
     }
 
     @Override
     public void createRecord(String country, Record record, CryptoManager cryptoManager) throws StorageClientException, StorageCryptoException, StorageServerException {
-        String url = getEndpoint(concatUrl(country), country);
-        agent.request(url, URI_POST, JsonUtils.toJsonString(record, cryptoManager), ApiResponse.WRITE);
+        String audienceUrl = getEndpoint(country);
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/dao/impl/HttpDaoImpl.java:122]: that should better be endpoint (var name). It's `audience url` only for oauth :)

File: src/main/java/com/incountry/residence/sdk/tools/dao/impl/HttpDaoImpl.java [hunk 3]
```diff
@@ -104,34 +107,38 @@ private String getEndpoint(String path, String country) throws StorageClientExce
...
+        String endpoint = audienceUrl + concatUrl(country);
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/dao/impl/HttpDaoImpl.java:123]: concatUrl is really a bad name for this method. It's name is misleading. We'd better refactor it somehow.
The same goes for createUrl method.
Let's have smth like:

- getRecordUrl(endpoint, country, key) -> for urls like `<endpoint>/v2/records/<country>/<key>`
- getRecordsUrl(endpoint, country, action) -> for urls like `<endpoint>/v2/records/<country>/<action>` (e.g. `v2/records/us/batchWrite`) 
- or combine them in a single method, like getPoPAPIUrl(endpoint, country, key, action)

File: src/test/java/com/incountry/residence/sdk/ProxyUtilsTest.java [hunk 4]
```diff
@@ -2,18 +2,19 @@
...
 
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
-public class ProxyUtilsTest {
+class ProxyUtilsTest {
 
     static class FakeHttpAgent implements HttpAgent {
         @Override
-        public String request(String endpoint, String method, String body, Map<Integer, ApiResponse> codeMap) {
+        public String request(String endpoint, String method, String body, Map<Integer, ApiResponse> codeMap, TokenClient tokenClient, String audienceUrl, int retryCount) {
```
Comment (RommeroInCountry) [inline src/test/java/com/incountry/residence/sdk/ProxyUtilsTest.java:17]: let's make namings more clear - audienceUrl -> popInstanceUrl (or popApiInstanceUrl). Let this url be audienceUrl only inside TokenClient

**Pull request merged**